packages/common/src/failure.ts

```diff
@@ -446,122 +327,320 @@ export function ensureTemporalFailure(err: unknown): TemporalFailure {
+          this.options.payloadConverter,
+          0,
+          failure.timeoutFailureInfo.lastHeartbeatDetails?.payloads
+        ),
+        failure.timeoutFailureInfo.timeoutType ?? TimeoutType.TIMEOUT_TYPE_UNSPECIFIED
+      );
+    }
+    if (failure.terminatedFailureInfo) {
+      return new TerminatedFailure(failure.message ?? undefined, this.optionalFailureToOptionalError(failure.cause));
+    }
+    if (failure.canceledFailureInfo) {
+      return new CancelledFailure(
+        failure.message ?? undefined,
+        arrayFromPayloads(this.options.payloadConverter, failure.canceledFailureInfo.details?.payloads),
+        this.optionalFailureToOptionalError(failure.cause)
+      );
+    }
+    if (failure.resetWorkflowFailureInfo) {
+      return new ApplicationFailure(
+        failure.message ?? undefined,
+        'ResetWorkflow',
+        false,
+        arrayFromPayloads(
+          this.options.payloadConverter,
+          failure.resetWorkflowFailureInfo.lastHeartbeatDetails?.payloads
+        ),
+        this.optionalFailureToOptionalError(failure.cause)
+      );
+    }
+    if (failure.childWorkflowExecutionFailureInfo) {
+      const { namespace, workflowType, workflowExecution, retryState } = failure.childWorkflowExecutionFailureInfo;
+      if (!(workflowType?.name && workflowExecution)) {
+        throw new TypeError('Missing attributes on childWorkflowExecutionFailureInfo');
+      }
+      return new ChildWorkflowFailure(
+        namespace ?? undefined,
+        workflowExecution,
+        workflowType.name,
+        retryState ?? RetryState.RETRY_STATE_UNSPECIFIED,
+        this.optionalFailureToOptionalError(failure.cause)
+      );
+    }
+    if (failure.activityFailureInfo) {
+      if (!failure.activityFailureInfo.activityType?.name) {
+        throw new TypeError('Missing activityType?.name on activityFailureInfo');
+      }
+      return new ActivityFailure(
+        failure.activityFailureInfo.activityType.name,
+        failure.activityFailureInfo.activityId ?? undefined,
+        failure.activityFailureInfo.retryState ?? RetryState.RETRY_STATE_UNSPECIFIED,
+        failure.activityFailureInfo.identity ?? undefined,
+        this.optionalFailureToOptionalError(failure.cause)
+      );
+    }
+    return new TemporalFailure(failure.message ?? undefined, this.optionalFailureToOptionalError(failure.cause));
   }
-  if (typeof error === 'string') {
-    return error;
+
+  failureToError(failure: ProtoFailure): TemporalFailure {
+    const err = this.failureToErrorInner(failure);
+    err.stack = failure.stackTrace ?? '';
+    err.failure = failure;
+    if (failure.encodedAttributes) {
+      const attrs = this.options.payloadConverter.fromPayload<DefaultEncodedFailureAttributes>(
+        failure.encodedAttributes
+      );
+      // Don't apply encodedAttributes unless they conform to an expected schema
+      if (typeof attrs === 'object' && attrs !== null) {
+        const { message, stack_trace } = attrs;
```

Comment:
Does this fail if our attributes are not present?

Reply:
It does not fail, it will fail if the payload converter cannot convert from payload though which I think is reasonable.

Reply:
:+1: My mistake, I assumed destructuring failed on missing key